sstables/sstables.cc

```diff
@@ -1559,6 +1559,21 @@ future<> sstable::reload_reclaimed_components() {
     sstlog.info("Reloaded bloom filter of {}", get_filename());
 }
 
+future<uint64_t> sstable::bytes_on_disk_uncached() const noexcept {
+    uint64_t bytes = 0;
+
+    co_await coroutine::parallel_for_each(_recognized_components, [this, &bytes] (const component_type& type) -> future<> {
+        try {
+            file f = co_await _storage->open_component(*this, type, open_flags::ro, file_open_options{}, false);
+            struct stat st = co_await f.stat().finally([&f] { return f.close(); });
```

Comment:
Wrinkles nose at coroutine/continuation mixes.

Is this really necessary? Can't we collect the size when we open the files (or when we append to them/delete them) rather that in a separate procedure?

Reply:
> Wrinkles nose at coroutine/continuation mixes.

Justifiedly so. I had four options (as far as I could tell):
- Don't close `f` explicitly, just let the destructor (?) deal with it. My own wrinkle factor: 3.
- Close `f` explicitly, using implementing the "finally" logic with explicit try / catch / std::current_exception. Wrote the code and cringed hard. Wrinkle factor: 3+.
- Browsed the source code and found prior art in (e.g.) `sstables/storage.cc`: `co_await with_closeable`. Generic, but hard to read. Wrinkle factor: 2.
- Figured that I might as well open-code `with_closeable` by chaining `.finally`, for such a short lambda. Wrinkle factor: 1 (not zero, but least annoying).

I can do `finally` with explicit try/catch/etc, but it was eye-poppingly awkard, for me. Please confirm if you want me to do that (or, of course, if I'm missing the obvious idiomatic solution, please let me know that one).

> Is this really necessary? Can't we collect the size when we open the files (or when we append to them/delete them) rather that in a separate procedure?

This is the question I've been asking since first looking at #18191. One of the first things I asked for was a design document or some kind of state diagram that described the state transitions / calling order of APIs. If we know what happens after what, then we can set up new data dependencies, i.e., precompute the disk usage and then just return that value. Thus far nobody shared any insight about the state transitions; so I'm getting the impression that the APIs and their interdependencies are more or less ad-hoc.

Metadata size may already be calculated in advance, but I don't understand what (and when) decides whether the sst will have an index file and/or a data file. In particular, if you look at commit 2dbae856f84e, `bytes_on_disk` was changed to require all three to be nonzero (metadata, index, data). This requirement is not satisfied however when scylla starts up with an otherwise empty database, and compacts some system-level sstables. Commit 2dbae856f84e didn't explain why requiring all three kinds of data for disk size calculation was justified.

In summary, I calculated the disk usage like this in v1 _precisely_ in order to force the question out in the open. The proposed "always re-stat" approach sucks, but I don't see the alternative, as long as we can't exactly specify the state transitions / API call order.

(Sorry about the wall of text.)

Reply:
I'll use `co_await with_closeable` in v2.